### PR TITLE
Add defaults during concat 508

### PR DESCRIPTION
**Pull request by scottcha**
- [x] Closes #508 
 - [x] Tests added
 - [x] Passes `black . && mypy . && flake8`
 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API

Contined on issue #508 by removing exception when concat two datasets with disjoint variables and instead add the missing variable with np.nan.

### CODE REVIEW

File: doc/whats-new.rst [hunk 1]
```diff
@@ -100,6 +100,8 @@ Bug fixes
   (:issue:`3402`). By `Deepak Cherian <https://github.com/dcherian/>`_
 - Allow appending datetime and bool data variables to zarr stores.
   (:issue:`3480`). By `Akihiro Matsukawa <https://github.com/amatsukawa/>`_.
+- Make :py:func:`~xarray.concat` more robust when merging variables present in some datasets but
```
Comment (dcherian) [inline doc/whats-new.rst]:
```suggestion
- Make :py:func:`~xarray.concat` more robust when concatenating variables present in some datasets but
```

File: xarray/core/concat.py [hunk 2]
```diff
@@ -370,10 +372,22 @@ def ensure_common_dims(vars):
...
-            except KeyError:
-                raise ValueError("%r is not present in all datasets." % k)
+            variables = []
+            for ds in datasets:
+                # if one of the variables doesn't exist find one which does
+                # and use it to create a fill value
+                if k not in ds.variables:
+                    for ds in datasets:
+                        if k in ds.variables:
+                            # found one to use as a fill value, fill with np.nan
```
[the remark this answers is not in the document]
Comment (scottcha) [reply xarray/core/concat.py]: Ahh, good point.  I overlooked that there was a fill_value parameter.

File: xarray/tests/test_combine.py [hunk 3]
```diff
@@ -742,10 +742,16 @@ def test_auto_combine(self):
...
             Dataset({"y": ("a", [1]), "x": ("a", [1])}),
         ]
+
         actual = auto_combine(objs)
         expected = Dataset({"x": ("a", [0, 1]), "y": ("a", [0, 1])})
         assert_identical(expected, actual)
 
+        objs = [Dataset({"x": [0], "y": [0]}), Dataset({"x": [0]})]
+        actual = auto_combine(objs)
+        expected = Dataset({"x": [0], "y": [0, np.nan]})
```
Comment (dcherian) [inline xarray/tests/test_combine.py]: This is interesting. I guess `y` is in `concat_over` so it gets filled. I wonder if filling in coordinate variables with NaNs is a good idea.

File: doc/whats-new.rst [hunk 4]
```diff
@@ -100,6 +100,8 @@ Bug fixes
   (:issue:`3402`). By `Deepak Cherian <https://github.com/dcherian/>`_
 - Allow appending datetime and bool data variables to zarr stores.
   (:issue:`3480`). By `Akihiro Matsukawa <https://github.com/amatsukawa/>`_.
+- Make :py:func:`~xarray.concat` more robust when concatenating variables present in some datasets but
+  not others (:issue:`508`). By `Scott Chamberlin <http://github.com/scottcha>`_.
```
Comment (keewis) [inline doc/whats-new.rst]:
```suggestion
  not others (:issue:`508`). By `Scott Chamberlin <https://github.com/scottcha>`_.
```
until we merge #3516, the some of the old entries don't have this format, but let's have all new entries match it.

File: xarray/core/concat.py [hunk 5]
```diff
@@ -370,10 +372,29 @@ def ensure_common_dims(vars):
...
-            try:
-                vars = ensure_common_dims([ds.variables[k] for ds in datasets])
-            except KeyError:
-                raise ValueError("%r is not present in all datasets." % k)
+            variables = []
+            for ds in datasets:
+                # if one of the variables doesn't exist find one which does
+                # and use it to create a fill value
+                if k not in ds.variables:
+                    for ds in datasets:
```
Comment (shoyer) [inline xarray/core/concat.py]: This nested loop through `datasets` concerns me here. It means that concat will run in quadratic time with respect to the number of datasets being concatenated. This probably make `xarray.concat` very slow on 1,000 datasets and outrageously slow on 10,000 datasets, both of which happen with some regularity.

it would be best to write this using a separate pass to create dummy versions of each Variable, which could be reused when appropriate.

File: xarray/core/concat.py [hunk 6]
```diff
@@ -370,10 +372,29 @@ def ensure_common_dims(vars):
...
+                            if fill_value is dtypes.NA:
+                                dtype, fill_value = dtypes.maybe_promote(
+                                    ds.variables[k].dtype
+                                )
+                            else:
+                                dtype = ds.variables[k].dtype
+
+                            filled = full_like(
+                                ds.variables[k], fill_value=fill_value, dtype=dtype
+                            )
```
Comment (shoyer) [inline xarray/core/concat.py]: I am concerned that this dummy variable may not always be the right size.

For example, supposing we are concatenating two Dataset along the existing dimension `'x'`. The first dataset has size `x=1` and the second has size `x=2`. If a variable is missing from one but not the other, the "dummy" variable would always have the wrong size, resulting in a total length of 2 or 4 but not 3.

To properly handle this, I think you will need to index out the concatenated dimension from the dummy variable (where-ever it is found), and then use `expand_dims` to add it back in the appropriate size for the current dataset.

File: xarray/tests/test_concat.py [hunk 7]
```diff
@@ -35,17 +35,28 @@ def test_concat_compat():
...
+        {
+            "has_x_y": (
+                ("q", "y", "x"),
+                [[[np.nan, np.nan], [3, 4]], [[1, 2], [np.nan, np.nan]]],
+            ),
+            "has_x": (("q", "x"), [[1, 2], [1, 2]]),
+            "no_x_y": (("q", "z"), [[1, 2], [1, 2]]),
+        },
+        coords={"x": [0, 1], "y": [0, 1], "z": [-1, -2], "q": [0, np.nan]},
+    )
```
Comment (shoyer) [inline xarray/tests/test_concat.py]: please add this in immediately above where it is used, i.e.,, down 5 lines from here

File: xarray/tests/test_combine.py (hunk 3 above)
Comment (shoyer) [reply xarray/tests/test_combine.py]: This doesn't look right to me. I think coordinates/indexes should be excluded from auto-filling with NaN.

**Comment by shoyer** [comment]
Thanks for working on this important issue!

There are a lot of edge cases that can come up in `concat`, so I think it would be very helpful to try to enumerate a broader set of unit tests for thoroughly testing this. For example:
- Pre-existing vs non-pre-existing dimension
- Pre-existing dimensions of different sizes
- Missing data variables vs coordinates vs indexed coordinates

File: xarray/core/concat.py (hunk 5 above)
Comment (dcherian) [reply xarray/core/concat.py]: > it would be best to write this using a separate pass to create dummy versions of each Variable, which could be reused when appropriate.

This could happen in `calc_concat_over`

File: xarray/tests/test_combine.py (hunk 3 above)
Comment (scottcha) [reply xarray/tests/test_combine.py]: Ok, I'll revert the change to the test and update the logic.

**Comment by scottcha** [comment]
Ok, I'll work on extending the updates with the feedback and additional tests.  Thanks!

File: xarray/core/concat.py [hunk 8]
```diff
@@ -370,10 +372,29 @@ def ensure_common_dims(vars):
...
+                if k not in ds.variables:
+                    for ds in datasets:
+                        if k in ds.variables:
+                            # found one to use as a fill value, fill with fill_value
+                            if fill_value is dtypes.NA:
+                                dtype, fill_value = dtypes.maybe_promote(
+                                    ds.variables[k].dtype
+                                )
+                            else:
+                                dtype = ds.variables[k].dtype
```
[the remark this answers is not in the document]
Comment (scottcha) [reply xarray/core/concat.py]: Ok this is in the new updated PR.

File: xarray/core/concat.py (hunk 6 above)
Comment (scottcha) [reply xarray/core/concat.py]: Ok, i'm not really sure I understand this case.  Any chance you can provide a test which I can use which would help?

File: xarray/core/concat.py (hunk 5 above)
Comment (scottcha) [reply xarray/core/concat.py]: The new PR contains improved logic but still required me to go through the list of data_sets a few times.  I think the new worst case runtime is O(D*N^2) where D is num of datasets and N is number of variables in final list.  If no fill value are required then it will be O(D*N).  
I did some perf testing with the new logic versus the old and I don't really see a significant difference but would love addition feedback if there is a better way.

Perf result for concat 720 files via open_mfdataset Parallel=False for PR:
58.7 s ± 143 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
Original result 
58.1 s ± 251 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

For 4359 files via open_mfdataset Parallel=False for PR:
5min 54s ± 840 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
sorry I don't really have a good real-world dataset this large w/out missing values to test the original implementation. But this dataset ~6x larger took ~6x more time even with the penalty to cache and fill the missing values.

I don't currently have good data without missing variables larger than this (hence the PR :) ) 

I was also not sure I should overload the logic in calc_concat_over to do more but I could re-review this if the logic in the new PR looks like it should be refactored that way.

**Comment by scottcha** [comment]
Hi, I've provided a new update to this PR (sorry it took me awhile both to get more familiar with the code and find the time to update the PR).  I improved the logic to be a bit more performant and handle more edge cases as well as updated the test suite.  I have a few questions:

1. The tests I wrote are a bit more verbose than the tests previously.  I can tighten them down but I found it was easier for me to read the logic in this form.  Please let me know what you prefer.
2. I'm still not quite sure I've captured all the scenarios as I'm a pretty basic xarray user so please let me know if there is still something I'm missing.

**Comment by shoyer** [comment]
I'll take a look at this more carefully soon. But I do think it is a hard
requirement that concat runs in linear time (with respect to the total
number of variables across all datasets).

**Comment by kmuehlbauer** [comment]
@scottcha If found this while searching. Have the same requirements, means missing DataArrays in some Datasets of a timeseries to be concatenated. I've already some hacks and workarounds in place for my specific use cases, but it would be really great if this could be handled by xarray.

I'll try to test your current implementation against my source data and will report my findings here.

**Comment by kmuehlbauer** [comment]
@scottcha @shoyer For one of my use cases (240 datasets, 1 with missing variables) I do not see any performance penalties using this implementation compared to the current. But this might be due to the fact, that the most time consuming part is the `expand_dims` for every dataset, which accounts for roughly 80% overall concat runtime.

If I can be of any help to push this over the line, please ping me.

**Comment by dcherian** [comment]
>  the most time consuming part is the expand_dims for every dataset, which accounts for roughly 80% overall concat runtime.

Hmmm... maybe we need a short-circuit version of `expand_dims`?

**Comment by dcherian** [comment]
>  the most time consuming part is the expand_dims for every dataset, which accounts for roughly 80% overall concat runtime.

Hmmm... maybe we need a short-circuit version of `expand_dims`?

**Comment by kmuehlbauer** [comment]
@dcherian Just to clarify, the concatenation is done along a new dimension (which has to be created by expand_dims). What do you mean by short-clrcuit in this context?

**Comment by scottcha** [comment]
@kmuehlbauer @dcherian @shoyer  If it would be easier it could abandon this PR and resubmit a new one as the code has drastically changed since the original comments were provided?

**Comment by shoyer** [comment]
Can you explain why you think you need the nested iteration over dataset variables? What ordering are you trying to achieve?

File: xarray/core/concat.py [hunk 9]
```diff
@@ -366,25 +370,101 @@ def ensure_common_dims(vars):
...
+            if var_list[i] not in union_of_variables:
+                # need to determine the correct place
+                # first add the new item which will be at the end
+                union_of_variables[var_list[i]] = None
+                union_of_variables.move_to_end(var_list[i])
+                # move any items after this in the variables list to the end
+                # this will only happen for missing variables
+                for j in range(i + 1, len(var_list)):
+                    if var_list[j] in union_of_variables:
+                        union_of_variables.move_to_end(var_list[j])
```
Comment (scottcha) [inline xarray/core/concat.py]: @shoyer  if this is the code you are referring to this have two purposes:

1. Find a complete set of variables even if the first datatset in the concat list has a missing variable (the previous implementation assums the first datatset has all variables)
2. Maintains the order of those variables (which is essentially the sorting operations happening when a missing variable is encountered) which was documented as a requirement for groupby in the previous implementation.

Comment (shoyer) [reply xarray/core/concat.py]: I'm not sure if preserving ordering is really essential, though I guess it would be nice to have.

The fundamental problem here is efficiently determining a consistent merge order between lists. This is pretty similar to some code I once wrote in TensorFlow. It only handles merging two lists efficiently, but hopefully is a good model. The fundamental idea is to simultaneously consume elements across all the lists at once.

I think there is no reason why it could not be extended to N-lists (though it would also need to be changed to fall-back to order of appearance rather than raising an error):
https://github.com/tensorflow/tensorflow/blob/v1.15.0/tensorflow/contrib/labeled_tensor/python/ops/core.py#L919

Either way, the logic should definitely live in a separate helper function, which makes it easier to test.

Comment (scottcha) [reply xarray/core/concat.py]: Regarding ordering I was going off the previous comment which said 
`    # stack up each variable to fill-out the dataset (in order)
    # n.b. this loop preserves variable order, needed for groupby.`
I believe one of the groupby test was also checking for that but I can't really recall at this point (regardless all existing groupby tests are currently passing in the pr).

I liked the code you linked and took a little detour to to try to incorporate it in to my PR.  I pulled it back out once I realized two things:

1. The conversion of the input list to a set I thought was a bit risky since the order isn't necessarily guaranteed (esp < python 3.7 where dicts weren't ordered by default) and it why my implementation was relying on ordered dicts.
2. The case where no consistent ordering was possible returned None while I didn't necessarily think that was appropriate for this code since there isn't really a strict necessity for variable ordering and I'm not sure you want to go deeper down that path.

I did spend a bit of time trying to write the generalized n dimension version of the consistent_ordering code but it was getting quite complex and was potentially hiding some complexity under some syntactic sugar.  I ended up refactoring the piece of code in question to an internal method (as its still fairly tied to the implementation of the public method) and put a note that its a potential candidate for a refactor.

The PR is updated with these changes.

Comment (kmuehlbauer) [reply xarray/core/concat.py]: @scottcha @shoyer I've tested again the different approaches. If there are only occasional misses I workes quite well. But in corner cases (two neighboring variables miss in consecutive datasets) it can have unwanted results. I'll add some code after the weekend.

From what I read this problem is closely related to the shortest common supersequence problem. I've checked on the implementations and it works very well in terms of result, but  is (currently) quite slow.

There should be some checks to possibly find one Dataset which contains all variables and can be used for output sorting. If none such is available then... 

If there is a correct solution possible, the code should find it. Just my 2c.

Comment (scottcha) [reply xarray/core/concat.py]: You are right that this is a special case of shortest common supersequence though since there shouldn't be repeated values in any sequence it might be easier to solve.

@kmuehlbauer  can you provide a case where you think the ordering determined by the current algorighm isn't providing the expected results?  I tried update the PR with a test case for the multiple neighboring missing variables (as well as explicit asserts on the data_var ordering) and I'm still getting expected results.  It would be great to see what you observed.

It may be time to actually ask what you want the behavior to really be in this case before introducing additional complexity.  I just read through some of the pandas issues and looked like they dealt with this as well https://github.com/pandas-dev/pandas/issues/4588.  Is that the behavior you would like in xarray?  I like the alignment at least for the default behavior with the sql behavior but I think its really up to the xarray owners?

Comment (kmuehlbauer) [reply xarray/core/concat.py]: @scottcha This is from the top of my head, so bear with me, if this isn't creating the unwanted effects.

```python
ds1 = ['d1', 'd3' , 'd4' , 'd5' , 'd6' ] 
ds2 = ['d1', 'd2' , 'd4' , 'd5' , 'd6' ] 
ds3 = ['d1', 'd2' , 'd3' , 'd5' , 'd6' ] 
ds4 = ['d1', 'd2' , 'd3' , 'd4' , 'd6' ] 
```

This is an example where one variable is missing in each Dataset, but the correct ordering is obvious. I hope I got it right. If not, I have to look it up on Monday earliest.

I'll test your additions/changes next week, currently travelling.

**Comment by kmuehlbauer** [comment]
@scottcha @shoyer below is a minimal example where one variable is missing in each file.

```python
import random
random.seed(123)
random.randint(0, 10)

# create var names list with one missing value
orig = [f'd{i:02}' for i in range(10)]
datasets = []
for i in range(1, 9):
    l1 = orig.copy()
    l1.remove(f'd{i:02}')
    datasets.append(l1)

# create files
for i, dsl in enumerate(datasets):
    foo_data = np.arange(24).reshape(2, 3, 4)
    with nc.Dataset(f'test{i:02}.nc', 'w') as ds:
        ds.createDimension('x', size=2)
        ds.createDimension('y', size=3)
        ds.createDimension('z', size=4)
        for k in dsl:
            ds.createVariable(k, int, ('x', 'y', 'z'))
            ds.variables[k][:] = foo_data

flist = glob.glob('test*.nc')
dslist = []
for f in flist:
    dslist.append(xr.open_dataset(f))

ds2 = xr.concat(dslist, dim='time')
ds2
```
Output: 

```
<xarray.Dataset>
Dimensions:  (time: 8, x: 2, y: 3, z: 4)
Dimensions without coordinates: time, x, y, z
Data variables:
    d01      (x, y, z) int64 0 1 2 3 4 5 6 7 8 9 ... 15 16 17 18 19 20 21 22 23
    d00      (time, x, y, z) int64 0 1 2 3 4 5 6 7 8 ... 16 17 18 19 20 21 22 23
    d02      (time, x, y, z) float64 0.0 1.0 2.0 3.0 4.0 ... 20.0 21.0 22.0 23.0
    d03      (time, x, y, z) float64 0.0 1.0 2.0 3.0 4.0 ... 20.0 21.0 22.0 23.0
    d04      (time, x, y, z) float64 0.0 1.0 2.0 3.0 4.0 ... 20.0 21.0 22.0 23.0
    d05      (time, x, y, z) float64 0.0 1.0 2.0 3.0 4.0 ... 20.0 21.0 22.0 23.0
    d06      (time, x, y, z) float64 0.0 1.0 2.0 3.0 4.0 ... 20.0 21.0 22.0 23.0
    d07      (time, x, y, z) float64 0.0 1.0 2.0 3.0 4.0 ... 20.0 21.0 22.0 23.0
    d08      (time, x, y, z) float64 0.0 1.0 2.0 3.0 4.0 ... nan nan nan nan nan
    d09      (time, x, y, z) int64 0 1 2 3 4 5 6 7 8 ... 16 17 18 19 20 21 22 23
```

Three cases here:

- `d00` and `d09` are available in all datasets, and they are concatenated correctly (keeping dtype)
- `d02` to `d08` are missing in one dataset and are filled with the created dummy variable, but the dtype is converted to float64
- `d01` is not handled properly, because it is missing in the first dataset, this is due to checking only variables of first dataset in [`_calc_concat_over`](https://github.com/scottcha/xarray/blob/cf5b8bdfb0fdaf626ecf7b83590f15aa9aef1d6b/xarray/core/concat.py#L235)

```python
            elif opt == "all":
                concat_over.update(
                    set(getattr(datasets[0], subset)) - set(datasets[0].dims)
                )
```
    and from putting `d01` in [`result_vars`](https://github.com/scottcha/xarray/blob/cf5b8bdfb0fdaf626ecf7b83590f15aa9aef1d6b/xarray/core/concat.py#L329) before iterating to find missing variables.

File: xarray/core/concat.py [hunk 10]
```diff
@@ -366,25 +370,89 @@ def ensure_common_dims(vars):
...
+            if variable_key not in ds.variables:
+                continue
+
+            v_fill_value = fill_value
+            dtype, v_fill_value = dtypes.get_fill_value_for_variable(
+                ds[variable_key], fill_value
+            )
+
+            union_of_variables[variable_key] = full_like(
+                ds[variable_key], fill_value=v_fill_value, dtype=dtype
```
Comment (kmuehlbauer) [inline xarray/core/concat.py]: This need to be `ds.variable[variable_key], fill_value=v_fill_value, dtype=dtype`, otherwise it will fail later (DataArray has no `set_dim`)

File: xarray/core/concat.py [hunk 11]
```diff
@@ -1,7 +1,9 @@
 import pandas as pd
+from collections import OrderedDict
```
[the remark this answers is not in the document]
Comment (scottcha) [reply xarray/core/concat.py]: Ok, I didn't realize that it was 3.6+ only.  Will change to dict.

File: xarray/tests/test_concat.py [hunk 12]
```diff
@@ -44,10 +110,438 @@ def test_concat_compat():
...
+
+    with raises_regex(ValueError, "coordinates in some datasets but not others"):
         concat([ds2, ds1], dim="q")
 
 
+def test_concat_missing_var():
+    datasets = create_concat_datasets(2, 123)
+    vars_to_drop = ["humidity", "precipitation", "cloud cover"]
+    datasets[0] = datasets[0].drop_vars(vars_to_drop)
+    datasets[1] = datasets[1].drop_vars(vars_to_drop + ["pressure"])
```
[the remark this answers is not in the document]
Comment (scottcha) [reply xarray/tests/test_concat.py]: I'll submit an update with the suggested changes.  I agree that I'm not sure where order should matter as long as the result is deterministic.

File: xarray/core/concat.py (hunk 10 above)
Comment (scottcha) [reply xarray/core/concat.py]: Thanks for the feedback and the above test.  I'll try to incorporate your suggested test as well as the rest of the pending comments in the next update.

**Comment by dcherian** [comment]
I am now wondering if we can use `align` to do the filling for us.

Example: goal is concat along 'x' with result dataset having `x=[1,2,3,4]`
1. Loop through datasets and assign coordinate values as appropriate. 
2. Break datasets up into mappings `collected = {"variable": [var1_at_x=1, var2_at_x=2, var4_at_x=4]}` -> there's some stuff in `merge.py` that could be reused for this
3. concatenate these lists to get a new mapping `concatenated = {"variable": [var_at_x=[1,2,4]]}`
4. apply `reindexed = {concatenated[var].reindex(x=[1,2,3,4], fill_value=...) for var in concatenated}`
5. create dataset `Dataset(reindexed)`

Step 1 would be where we deal with all the edge cases mentioned in @shoyer's comment viz

> For example:
> 
> Pre-existing vs non-pre-existing dimension
> Pre-existing dimensions of different sizes
> Missing data variables vs coordinates vs indexed coordinates

**Comment by scottcha** [comment]
I just pushed an incomplete set of changes as @kmuehlbauer tests have demonstrated there was some incomplete cases the PR still isn't handling.  
Here is a summary:
1. I've simplified the logic based on @dcherian comments but in order to keep the result deterministic needed to use list logic instead of set logic.  I also kept the OrderedDict instead of going with the default dict as the built in ordering methods as of py 3.6 were still insufficient for keeping the ordering consistent (I needed to pop FIFO) which doesn't seem possible until py 3.8.
2. I did add a failing test to capture the cases @kmuehlbauer  pointed out.

I'm not sure I have my head wrapped around xarray enough to address @dcherian's latest comments though which is why i'm sharing the code at this point.  All tests are passing except the new cases which were pointed out.

I'll try to continue to get time to update this but wanted to at least provide this status update at this point as its been awhile.

**Comment by mullenkamp** [comment]
Has this been implemented? Or is it still failing the tests?

**Comment by lukelbd** [comment]
Cool PR - looks like it's stale? Maybe someone should copy the work to a new one? Have been coming across this issue a lot in my work recently.

**Comment by kmuehlbauer** [comment]
@scottcha Are you still around and interested to bring this along? If not I could try to dive again into this.

**Comment by scottcha** [comment]
I'm still along and yes I do still need this functionality (I still sync back to this PR when I have data missing vars).  The issue was that the technical requirements got beyond what I was able to account for with the time I had available.  If you or someone else was interested in picking it up I'd be happy to evaluate against my use cases.

**Comment by keewis** [comment]
> I can try to rebase the PR onto latest main

I did try that a few months ago, but a lot has changed since the PR was opened so it might actually be easier to reimplement the PR?

**Comment by kmuehlbauer** [comment]
Thanks @keewis for the heads up. I'll have a look and if things get too complicated a reimplementation might be our best option.

**Comment by kmuehlbauer** [comment]
@scottcha @keewis I've tried hard, but finally decided to start from scratch, see #7400.
